From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Asriel Dreemurr <asriel.danctnix@gmail.com>
Date: Sun, 18 Oct 2020 00:31:45 +0700
Subject: [PATCH] (halium) init: start inside LXC container without SELinux

amartinz: keep first stage init setup for recovery, otherwise
recovery will not boot.

Signed-off-by: Alexander Martinz <alex@amartinz.at>
Signed-off-by: Eugenio Paolantonio (g7) <me@medesimo.eu>
---
 init/Android.bp           |  2 ++
 init/first_stage_init.cpp | 47 +++++++++++++++++++++++++++++----------
 init/init.cpp             |  8 ++++++-
 init/property_service.cpp | 24 ++++++++++++--------
 init/service.cpp          | 10 +++++----
 init/subcontext.cpp       |  3 ++-
 init/util.cpp             | 18 +++++++++------
 7 files changed, 78 insertions(+), 34 deletions(-)

diff --git a/init/Android.bp b/init/Android.bp
index 00ac6a752..fc55ee4aa 100644
--- a/init/Android.bp
+++ b/init/Android.bp
@@ -119,6 +119,7 @@ libinit_cc_defaults {
         "-Wall",
         "-Wextra",
         "-Wno-unused-parameter",
+        "-Wno-unused-function",
         "-Werror",
         "-Wthread-safety",
         "-DALLOW_FIRST_STAGE_CONSOLE=0",
@@ -503,6 +504,7 @@ cc_binary {
         "-Wall",
         "-Wextra",
         "-Wno-unused-parameter",
+        "-Wno-unused-function",
         "-Werror",
     ],
     static_libs: [
diff --git a/init/first_stage_init.cpp b/init/first_stage_init.cpp
index c7b7b0c13..44b8747a9 100644
--- a/init/first_stage_init.cpp
+++ b/init/first_stage_init.cpp
@@ -198,14 +198,23 @@ int FirstStageMain(int argc, char** argv) {
     CHECKCALL(setenv("PATH", _PATH_DEFPATH, 1));
     // Get the basic filesystem setup we need put together in the initramdisk
     // on / and then we'll let the rc file figure out the rest.
-    CHECKCALL(mount("tmpfs", "/dev", "tmpfs", MS_NOSUID, "mode=0755"));
-    CHECKCALL(mkdir("/dev/pts", 0755));
-    CHECKCALL(mkdir("/dev/socket", 0755));
+
+    // We need to disable certain things for Anbox while the recovery needs it.
+    bool is_recovery = IsRecoveryMode();
+
+    // Disabled for Anbox, mounted by host system instead
+    if (!is_recovery) {
+        mkdir("/dev/socket", 0755);
+    } else {
+        CHECKCALL(mount("tmpfs", "/dev", "tmpfs", MS_NOSUID, "mode=0755"));
+        CHECKCALL(mkdir("/dev/pts", 0755));
+        CHECKCALL(mkdir("/dev/socket", 0755));
     CHECKCALL(mkdir("/dev/dm-user", 0755));
-    CHECKCALL(mount("devpts", "/dev/pts", "devpts", 0, NULL));
+        CHECKCALL(mount("devpts", "/dev/pts", "devpts", 0, NULL));
 #define MAKE_STR(x) __STRING(x)
-    CHECKCALL(mount("proc", "/proc", "proc", 0, "hidepid=2,gid=" MAKE_STR(AID_READPROC)));
+        CHECKCALL(mount("proc", "/proc", "proc", 0, "hidepid=2,gid=" MAKE_STR(AID_READPROC)));
 #undef MAKE_STR
+    }
     // Don't expose the raw commandline to unprivileged processes.
     CHECKCALL(chmod("/proc/cmdline", 0440));
     std::string cmdline;
@@ -216,8 +225,11 @@ int FirstStageMain(int argc, char** argv) {
     android::base::ReadFileToString("/proc/bootconfig", &bootconfig);
     gid_t groups[] = {AID_READPROC};
     CHECKCALL(setgroups(arraysize(groups), groups));
-    CHECKCALL(mount("sysfs", "/sys", "sysfs", 0, NULL));
-    CHECKCALL(mount("selinuxfs", "/sys/fs/selinux", "selinuxfs", 0, NULL));
+    // Disabled for Anbox
+    if (is_recovery) {
+        CHECKCALL(mount("sysfs", "/sys", "sysfs", 0, NULL));
+        CHECKCALL(mount("selinuxfs", "/sys/fs/selinux", "selinuxfs", 0, NULL));
+    }
 
     CHECKCALL(mknod("/dev/kmsg", S_IFCHR | 0600, makedev(1, 11)));
 
@@ -229,8 +241,14 @@ int FirstStageMain(int argc, char** argv) {
     CHECKCALL(mknod("/dev/urandom", S_IFCHR | 0666, makedev(1, 9)));
 
     // This is needed for log wrapper, which gets called before ueventd runs.
-    CHECKCALL(mknod("/dev/ptmx", S_IFCHR | 0666, makedev(5, 2)));
-    CHECKCALL(mknod("/dev/null", S_IFCHR | 0666, makedev(1, 3)));
+    // Can be created by LXC on Anbox
+    if (is_recovery) {
+        CHECKCALL(mknod("/dev/ptmx", S_IFCHR | 0666, makedev(5, 2)));
+        CHECKCALL(mknod("/dev/null", S_IFCHR | 0666, makedev(1, 3)));
+    } else {
+        mknod("/dev/ptmx", S_IFCHR | 0666, makedev(5, 2));
+        mknod("/dev/null", S_IFCHR | 0666, makedev(1, 3));
+    }
 
     // These below mounts are done in first stage init so that first stage mount can mount
     // subdirectories of /mnt/{vendor,product}/.  Other mounts, not required by first stage mount,
@@ -265,7 +283,8 @@ int FirstStageMain(int argc, char** argv) {
         for (const auto& [error_string, error_errno] : errors) {
             LOG(ERROR) << error_string << " " << strerror(error_errno);
         }
-        LOG(FATAL) << "Init encountered errors starting first stage, aborting";
+        // Halium: log errors but do not abort
+        // LOG(FATAL) << "Init encountered errors starting first stage, aborting";
     }
 
     LOG(INFO) << "init first stage started!";
@@ -357,7 +376,8 @@ int FirstStageMain(int argc, char** argv) {
         SwitchRoot("/first_stage_ramdisk");
     }
 
-    if (!DoFirstStageMount(!created_devices)) {
+    // Disabled for Anbox
+    if (is_recovery && !DoFirstStageMount()) {
         LOG(FATAL) << "Failed to mount required partitions early ...";
     }
 
@@ -376,8 +396,11 @@ int FirstStageMain(int argc, char** argv) {
     setenv(kEnvFirstStageStartedAt, std::to_string(start_time.time_since_epoch().count()).c_str(),
            1);
 
+
     const char* path = "/system/bin/init";
-    const char* args[] = {path, "selinux_setup", nullptr};
+    // Halium: skip selinux_setup
+    // const char* args[] = {path, "selinux_setup", nullptr};
+    const char* args[] = {path, "second_stage", nullptr};
     auto fd = open("/dev/kmsg", O_WRONLY | O_CLOEXEC);
     dup2(fd, STDOUT_FILENO);
     dup2(fd, STDERR_FILENO);
diff --git a/init/init.cpp b/init/init.cpp
index 942feb939..5646b69c2 100644
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -872,7 +872,10 @@ int SecondStageMain(int argc, char** argv) {
     // Queue an action that waits for coldboot done so we know ueventd has set up all of /dev...
     am.QueueBuiltinAction(wait_for_coldboot_done_action, "wait_for_coldboot_done");
     // ... so that we can start queuing up actions that require stuff from /dev.
-    am.QueueBuiltinAction(SetMmapRndBitsAction, "SetMmapRndBits");
+    // Disabled for Anbox
+    //am.QueueBuiltinAction(MixHwrngIntoLinuxRngAction, "MixHwrngIntoLinuxRng");
+    //am.QueueBuiltinAction(SetMmapRndBitsAction, "SetMmapRndBits");
+
     Keychords keychords;
     am.QueueBuiltinAction(
             [&epoll, &keychords](const BuiltinArguments& args) -> Result<void> {
@@ -887,6 +890,9 @@ int SecondStageMain(int argc, char** argv) {
     // Trigger all the boot actions to get us started.
     am.QueueEventTrigger("init");
 
+    // Disabled for Anbox
+    // am.QueueBuiltinAction(MixHwrngIntoLinuxRngAction, "MixHwrngIntoLinuxRng");
+
     // Don't mount filesystems or start core system services in charger mode.
     std::string bootmode = GetProperty("ro.bootmode", "");
     if (bootmode == "charger") {
diff --git a/init/property_service.cpp b/init/property_service.cpp
index 5a8fc5418..5e9d8d05f 100644
--- a/init/property_service.cpp
+++ b/init/property_service.cpp
@@ -142,7 +142,8 @@ void StopSendingMessages() {
 }
 
 bool CanReadProperty(const std::string& source_context, const std::string& name) {
-    const char* target_context = nullptr;
+    // Disabled for Anbox
+    /*const char* target_context = nullptr;
     property_info_area->GetPropertyInfo(name.c_str(), &target_context, nullptr);
 
     PropertyAuditData audit_data;
@@ -153,12 +154,14 @@ bool CanReadProperty(const std::string& source_context, const std::string& name)
     audit_data.cr = &cr;
 
     return selinux_check_access(source_context.c_str(), target_context, "file", "read",
-                                &audit_data) == 0;
+                                &audit_data) == 0;*/
+    return true;
 }
 
 static bool CheckMacPerms(const std::string& name, const char* target_context,
                           const char* source_context, const ucred& cr) {
-    if (!target_context || !source_context) {
+    // Disabled for Anbox
+    /*if (!target_context || !source_context) {
         return false;
     }
 
@@ -170,7 +173,8 @@ static bool CheckMacPerms(const std::string& name, const char* target_context,
     bool has_access = (selinux_check_access(source_context, target_context, "property_service",
                                             "set", &audit_data) == 0);
 
-    return has_access;
+    return has_access;*/
+    return true;
 }
 
 static uint32_t PropertySet(const std::string& name, const std::string& value, std::string* error) {
@@ -564,11 +568,12 @@ static void handle_property_set_fd() {
         prop_name[PROP_NAME_MAX-1] = 0;
         prop_value[PROP_VALUE_MAX-1] = 0;
 
-        std::string source_context;
-        if (!socket.GetSourceContext(&source_context)) {
+        std::string source_context = "";
+        // Disabled for Anbox
+        /*if (!socket.GetSourceContext(&source_context)) {
             PLOG(ERROR) << "Unable to set property '" << prop_name << "': getpeercon() failed";
             return;
-        }
+        }*/
 
         const auto& cr = socket.cred();
         std::string error;
@@ -593,11 +598,12 @@ static void handle_property_set_fd() {
         }
 
         std::string source_context;
-        if (!socket.GetSourceContext(&source_context)) {
+        // Disabled for Anbox
+        /*if (!socket.GetSourceContext(&source_context)) {
             PLOG(ERROR) << "Unable to set property '" << name << "': getpeercon() failed";
             socket.SendUint32(PROP_ERROR_PERMISSION_DENIED);
             return;
-        }
+        }*/
 
         const auto& cr = socket.cred();
         std::string error;
diff --git a/init/service.cpp b/init/service.cpp
index dcb3d0c3c..193b718bd 100644
--- a/init/service.cpp
+++ b/init/service.cpp
@@ -239,9 +239,10 @@ void Service::SetProcessAttributesAndCaps() {
     }
 
     if (!seclabel_.empty()) {
-        if (setexeccon(seclabel_.c_str()) < 0) {
+        // Disabled for Anbox
+        /*if (setexeccon(seclabel_.c_str()) < 0) {
             PLOG(FATAL) << "cannot setexeccon('" << seclabel_ << "') for " << name_;
-        }
+        }*/
     }
 
     if (capabilities_) {
@@ -453,7 +454,8 @@ Result<void> Service::Start() {
     }
 
     std::string scon;
-    if (!seclabel_.empty()) {
+    // Disabled for Anbox
+    /*if (!seclabel_.empty()) {
         scon = seclabel_;
     } else {
         auto result = ComputeContextFromExecutable(args_[0]);
@@ -461,7 +463,7 @@ Result<void> Service::Start() {
             return result.error();
         }
         scon = *result;
-    }
+    }*/
 
     // APEXd is always started in the "current" namespace because it is the process to set up
     // the current namespace.
diff --git a/init/subcontext.cpp b/init/subcontext.cpp
index fa48beab3..07258c17a 100644
--- a/init/subcontext.cpp
+++ b/init/subcontext.cpp
@@ -213,7 +213,8 @@ void Subcontext::Fork() {
 
         // We don't switch contexts if we're running the unit tests.  We don't use std::optional,
         // since we still need a real context string to pass to the builtin functions.
-        if (context_ != kTestContext) {
+        // Disabled in LXC container
+        if (false /*context_ != kTestContext*/) {
             if (setexeccon(context_.c_str()) < 0) {
                 PLOG(FATAL) << "Could not set execcon for '" << context_ << "'";
             }
diff --git a/init/util.cpp b/init/util.cpp
index 9f7bfdb5b..bff963876 100644
--- a/init/util.cpp
+++ b/init/util.cpp
@@ -88,18 +88,20 @@ Result<uid_t> DecodeUid(const std::string& name) {
  */
 Result<int> CreateSocket(const std::string& name, int type, bool passcred, mode_t perm, uid_t uid,
                          gid_t gid, const std::string& socketcon) {
-    if (!socketcon.empty()) {
+    // Disabled for Anbox
+    /*if (!socketcon.empty()) {
         if (setsockcreatecon(socketcon.c_str()) == -1) {
             return ErrnoError() << "setsockcreatecon(\"" << socketcon << "\") failed";
         }
-    }
+    }*/
 
     android::base::unique_fd fd(socket(PF_UNIX, type, 0));
     if (fd < 0) {
         return ErrnoError() << "Failed to open socket '" << name << "'";
     }
 
-    if (!socketcon.empty()) setsockcreatecon(nullptr);
+    // Disabled for Anbox
+    /*if (!socketcon.empty()) setsockcreatecon(nullptr);*/
 
     struct sockaddr_un addr;
     memset(&addr, 0 , sizeof(addr));
@@ -110,10 +112,11 @@ Result<int> CreateSocket(const std::string& name, int type, bool passcred, mode_
         return ErrnoError() << "Failed to unlink old socket '" << name << "'";
     }
 
-    std::string secontext;
+    // Disabled for Anbox
+    /*std::string secontext;
     if (SelabelLookupFileContext(addr.sun_path, S_IFSOCK, &secontext) && !secontext.empty()) {
         setfscreatecon(secontext.c_str());
-    }
+    }*/
 
     if (passcred) {
         int on = 1;
@@ -125,9 +128,10 @@ Result<int> CreateSocket(const std::string& name, int type, bool passcred, mode_
     int ret = bind(fd, (struct sockaddr *) &addr, sizeof (addr));
     int savederrno = errno;
 
-    if (!secontext.empty()) {
+    // Disabled for Anbox
+    /*if (!secontext.empty()) {
         setfscreatecon(nullptr);
-    }
+    }*/
 
     auto guard = android::base::make_scope_guard([&addr] { unlink(addr.sun_path); });
 
-- 
2.25.1

